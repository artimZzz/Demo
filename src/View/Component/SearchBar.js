import React, { Component, PureComponent } from 'react';
import { StyleSheet, ViewPropTypes, View, Animated, Easing, Text, TextInput } from 'react-native';
import { ifAndroid } from '../../Controller/Device/DeviceCheck';
import Ionicons from 'react-native-vector-icons/Ionicons';
import TouchableDebounce from './TouchableDebounce';
import PropTypes from 'prop-types';

export default class SearchBar extends PureComponent {

    /* constructor */
    constructor(props) {
        super(props);

        this.state = {
            isSearching: false,
            toggleAnim: new Animated.Value(0)
        };
        this.textInputRef = null;
        this.touchRef = null;

        /* callback */
        this.onPress = () => {
            this.state.toggleAnim.setValue(0);
            Animated.timing(
                this.state.toggleAnim,
                { toValue: 1.0, duration: 150, easing: Easing.linear }
            ).start(
                () => {
                    this.setState({ isSearching: true });
                }
            );

        }

        this.onBlur = () => {
            if (this.textInputRef._lastNativeText.length == 0) {
                this.setState(
                    { isSearching: false },
                    () => {
                        this.state.toggleAnim.setValue(1);
                        Animated.timing(
                            this.state.toggleAnim,
                            { toValue: 0, duration: 150, easing: Easing.linear }
                        ).start();
                    }
                );
            }
            if (this.props.onBlur) {
                this.props.onBlur();
            }
        }

        this.assignTouchRef = (ref) => this.touchRef = ref;

        this.assignRef = (ref, innerRef) => {
            this.textInputRef = ref;
            if (innerRef) {
                innerRef(ref);
            }
        }
    }

    /* render */
    render() {
        const { style, ...otherProps } = this.props;
        return (
            <View style={style}>
                {this.renderInner()}
            </View>
        );
    }

    renderInner = () => {
        const {
            forwardRef, forwardStyle, placeholder, onBlur, // props using
            style, ref, autoCapitalize, autoCorrect, autoFocus, spellCheck, // skip props
            ...otherProps // auto inherit props
        } = this.props;
        if (!this.state.isSearching) {
            let opacity = this.state.toggleAnim.interpolate(
                {
                    inputRange: [0, 1],
                    outputRange: [1, 0]
                }
            );
            let positionX = this.state.toggleAnim.interpolate(
                {
                    inputRange: [0, 1],
                    outputRange: [0, -100]
                }
            );
            return (
                <Animated.View style={[styles.searchCoverBox, { opacity: opacity, left: positionX }]}>
                    <TouchableDebounce ref={this.assignTouchRef} style={styles.searchCoverTouch} onPress={this.onPress}>
                        {this.renderSearchIcon()}
                        <Text style={styles.searchCoverText}>{placeholder}</Text>
                    </TouchableDebounce>
                </Animated.View>
            );
        }
        return (
            <View style={styles.searchInputBox}>
                {this.renderSearchIcon()}
                <TextInput
                    {...otherProps}
                    ref={(ref) => this.assignRef(ref, forwardRef)}
                    placeholder={placeholder}
                    autoCapitalize='none'
                    autoCorrect={false}
                    autoFocus={true}
                    spellCheck={false}
                    onBlur={this.onBlur}
                    style={[forwardStyle, styles.searchInput]}
                />
            </View>
        );
    }

    renderSearchIcon = () => {
        return <Ionicons name='ios-search' size={20} color='#00000066' />;
    }
}

SearchBar.propTypes = {
    style: ViewPropTypes.style,
    ref: PropTypes.func,
    forwardRef: PropTypes.func,
    forwardStyle: ViewPropTypes.style,
    placeholder: PropTypes.string
};
SearchBar.defaultProps = {
    style: {},
    ref: () => { },
    forwardRef: () => { },
    forwardStyle: {},
    placeholder: ''
};

const styles = StyleSheet.create(
    {
        searchCoverBox: {
            position: 'absolute', width: '100%', height: '100%'
        },
        searchCoverTouch: {
            width: '100%', height: '100%', alignItems: 'center', justifyContent: 'center', flexDirection: 'row'
        },
        searchCoverText: {
            marginLeft: 10, fontSize: 14, fontWeight: '400', color: '#00000066'
        },
        searchInputBox: {
            position: 'absolute', width: '100%', height: '100%', flexDirection: 'row', alignItems: 'center', paddingHorizontal: 10
        },
        searchInput: {
            flex: 1, height: '100%', marginLeft: 10, ...ifAndroid({ paddingBottom: 8 }, {})
        }
    }
);