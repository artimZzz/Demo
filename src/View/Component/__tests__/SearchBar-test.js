// __tests__/SearchBar-test.js
import React from 'react';
import SearchBar from '../SearchBar';

import renderer from 'react-test-renderer';

test('renders correctly', () => {
    const tree = renderer.create(<SearchBar />).toJSON();
    expect(tree).toMatchSnapshot();
});