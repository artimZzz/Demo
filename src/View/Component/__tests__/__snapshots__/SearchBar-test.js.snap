// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "height": "100%",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#00000066",
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        ï’¥
      </Text>
      <Text
        style={
          Object {
            "color": "#00000066",
            "fontSize": 14,
            "fontWeight": "400",
            "marginLeft": 10,
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
